core/trino-main/src/main/java/io/trino/operator/ExchangeClient.java

```diff
@@ -406,32 +313,21 @@ private synchronized void requestComplete(HttpPageBufferClient client)
     private synchronized void clientFinished(HttpPageBufferClient client)
     {
         requireNonNull(client, "client is null");
-        completedClients.add(client);
-        scheduleRequestIfNecessary();
-    }
-
-    private synchronized void clientFailed(Throwable cause)
-    {
-        // TODO: properly handle the failed vs closed state
-        // it is important not to treat failures as a successful close
-        if (!isClosed()) {
-            failure.compareAndSet(null, cause);
-            notifyBlockedCallers();
+        if (completedClients.add(client)) {
+            buffer.taskFinished(client.getRemoteTaskId());
         }
+        scheduleRequestIfNecessary();
     }
 
-    private boolean isFailed()
-    {
-        return failure.get() != null;
-    }
-
-    private void throwIfFailed()
+    private synchronized void clientFailed(HttpPageBufferClient client, Throwable cause)
     {
-        Throwable t = failure.get();
-        if (t != null) {
-            throwIfUnchecked(t);
-            throw new RuntimeException(t);
+        requireNonNull(client, "client is null");
+        if (completedClients.add(client)) {
+            buffer.taskFailed(client.getRemoteTaskId(), cause);
+            scheduler.execute(() -> taskFailureListener.onTaskFailed(client.getRemoteTaskId(), cause));
```

Comment:
nit: can you rename `scheduler` to `executorService`

Reply:
I guess the variable name was chosen to emphasize the executor is a scheduled executor. Scheduled executor is needed in `HttpPageBufferClient`. I would prefer to keep it as is.

Reply:
too many "schedulers" IMO. When I see "scheduler" I think stage or task scheduler. 
But whatever.

Reply:
Oh, I see your point. `scheduledExecutor`, wdyt? I can add a rename as a separate commit (rename variable in both `ExchangeClient` and `HttpPageBufferClient`)?

Reply:
Lgtm